### PR TITLE
[slang] Introduce net aliases bits duplication checks

### PR DESCRIPTION
**Pull request by likeamahoney**
Introduce 2 new net alias bit checks from SystemVerilog LRM section 10.11:

- check that net aliased to itself
- check that aliased bits specified more than once at module instance scope

### CODE REVIEW

**Comment by likeamahoney** [comment]
I think with it LRM section 10.11 is done

File: include/slang/ast/symbols/InstanceSymbols.h [hunk 1]
```diff
@@ -207,6 +208,10 @@ class SLANG_EXPORT InstanceBodySymbol : public Symbol, public Scope {
     const DefinitionSymbol& definition;
     mutable std::span<const Symbol* const> portList;
     std::span<const ParameterSymbolBase* const> parameters;
+    // Store any net alias bits to check it's possible intersections.
+    // Since alias statements can appear anywhere module instance statements can appears
+    // good option is to store aliased bits info in the module instance scope.
+    mutable AliasMap* aliasedBits;
```
Comment (MikePopoloski) [inline include/slang/ast/symbols/InstanceSymbols.h:214]: It seems unnecessary to have an alias map allocated for every instance. The rules about net alias overlap are global, so it seems fine to have just one alias map stored in the Compilation object that all aliases can access. Then you don't need a custom allocator for it either.

File: source/ast/symbols/MemberSymbols.cpp [hunk 2]
```diff
@@ -2196,7 +2196,6 @@ NetAliasSymbol& NetAliasSymbol::fromSyntax(const ASTContext& parentContext,
         if (!netType.isError()) {
             SmallVector<const IdentifierNameSyntax*> implicitNetNames;
             Expression::findPotentiallyImplicitNets(*expr, context, implicitNetNames);
-
```
Comment (MikePopoloski) [inline source/ast/symbols/MemberSymbols.cpp:2199]: Random newline deletion seems unrelated to the rest of the diff

File: source/ast/symbols/MemberSymbols.cpp [hunk 3]
```diff
@@ -2214,34 +2213,89 @@ struct NetAliasVisitor {
...
+                        aliasedBits[sym] = {
+                            AliasedBits{std::nullopt, std::nullopt, expr.sourceRange}};
+                    }
+                    break;
+                }
+                case ExpressionKind::ElementSelect: {
+                    if (const auto* sym = processExpr(expr)) {
+                        const auto& eSelExpr = expr.template as<ElementSelectExpression>();
+                        // If value can't be evaluated that means it is unbounded
+                        const auto evalVal = toInt(context.tryEval(eSelExpr.selector()));
```
Comment (MikePopoloski) [inline source/ast/symbols/MemberSymbols.cpp:2264]: I don't think this is correct in all cases. An example is when there is multiple selections of a multi-dimensional array. It's actually pretty tricky to figure out the statically assigned bits of an arbitrary expression -- that's essentially what the ValueDriver code in ValueSymbol.cpp is doing to be able to issue multi-driven errors. It occurs to me that it may make more sense to use that code to implement the net alias check. You could add a new flag to AssignFlags that says it's a net alias and do the check for overlap there.

File: include/slang/ast/symbols/InstanceSymbols.h (hunk 1 above)
Comment (likeamahoney) [reply include/slang/ast/symbols/InstanceSymbols.h]: reimplemented all in order to last commentary

File: source/ast/symbols/MemberSymbols.cpp (hunk 2 above)
Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: fixed

File: source/ast/ASTContext.cpp [hunk 4]
```diff
@@ -136,7 +136,8 @@ void ASTContext::setAttributes(const Expression& expr,
 
 void ASTContext::addDriver(const ValueSymbol& symbol, const Expression& longestStaticPrefix,
                            bitmask<AssignFlags> assignFlags) const {
-    if (flags.has(ASTFlags::NotADriver) || scope->isUninstantiated())
+    if ((flags.has(ASTFlags::NotADriver) && !assignFlags.has(AssignFlags::NetAlias)) ||
```
Comment (MikePopoloski) [inline source/ast/ASTContext.cpp:139]: I think you can get rid of this change if you simply don't use the NotADriver flag when binding the net alias expressions

File: source/ast/symbols/MemberSymbols.cpp [hunk 5]
```diff
@@ -2119,16 +2124,18 @@ std::span<const Expression* const> NetAliasSymbol::getNetReferences() const {
     auto syntax = getSyntax();
     SLANG_ASSERT(scope && syntax);
 
-    // TODO: there should be a global check somewhere that any given bit
-    // of a net isn't aliased to the same target signal bit multiple times.
     bitwidth_t bitWidth = 0;
     bool issuedError = false;
     SmallVector<const Expression*> buffer;
     ASTContext context(*scope, LookupLocation::after(*this),
                        ASTFlags::NonProcedural | ASTFlags::NotADriver);
```
Comment (MikePopoloski) [inline source/ast/symbols/MemberSymbols.cpp:2131]: Down here, you can remove the NotADriver

File: source/ast/symbols/MemberSymbols.cpp [hunk 6]
```diff
@@ -2119,16 +2124,18 @@ std::span<const Expression* const> NetAliasSymbol::getNetReferences() const {
...
     NetAliasVisitor visitor(context);
+    // In order to save the leftmost alias
+    SmallMap<const ValueSymbol*, const Expression*, 2> leftAlias;
+
+    const auto nets = syntax->as<NetAliasSyntax>().nets;
 
-    for (auto exprSyntax : syntax->as<NetAliasSyntax>().nets) {
+    for (auto exprSyntax : nets) {
         auto& netRef = Expression::bind(*exprSyntax, context);
         if (!netRef.requireLValue(context))
```
Comment (MikePopoloski) [inline source/ast/symbols/MemberSymbols.cpp:2140]: And then here pass your new AssignFlag and then you won't need the additional addDriver call below I think.

**Review by MikePopoloski — COMMENTED** [review]
Ok yeah, this is way better. I left some additional comments to help simplify this a bit more.

File: source/ast/symbols/MemberSymbols.cpp (hunk 6 above)
Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: done

File: source/ast/symbols/ValueSymbol.cpp [hunk 7]
```diff
@@ -7,6 +7,8 @@
 //------------------------------------------------------------------------------
 #include "slang/ast/symbols/ValueSymbol.h"
 
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (likeamahoney) [reply source/ast/symbols/ValueSymbol.cpp]: removed

File: source/ast/ASTContext.cpp (hunk 4 above)
Comment (likeamahoney) [reply source/ast/ASTContext.cpp]: fixed

File: source/ast/symbols/MemberSymbols.cpp (hunk 5 above)
Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp:2131]: The flag is needed to skip the first alias (the leftmost one) in order to prevent FP. I left it, but remove it after the first iteration of the loop.

File: source/ast/symbols/MemberSymbols.cpp [hunk 8]
```diff
@@ -2144,6 +2159,21 @@ std::span<const Expression* const> NetAliasSymbol::getNetReferences() const {
 
         netRef.visit(visitor);
         buffer.push_back(&netRef);
+
+        if (exprSyntax != *nets.begin()) {
```
Comment (MikePopoloski) [inline source/ast/symbols/MemberSymbols.cpp:2163]: This logic seems confusing to me; you really just need to check whether the same symbol appears in more than one of the net alias expressions right? The `aliasedSyms` map in the NetAliasVisitor is already tracking that; instead of insert_or_assign up above you can just check whether the insertion fails because the symbol already exists, in which case you can issue the error right there. Then you don't need any of this special casing for it here and the separate map tracking the same thing.

File: source/ast/symbols/MemberSymbols.cpp (hunk 5 above)
Comment (MikePopoloski) [reply source/ast/symbols/MemberSymbols.cpp:2131]: I'm not following why the first alias expression should be exempt from the duplicate checking. There isn't anything special about the first alias vs any others; you can reverse the order without changing the behavior.

I do think we need more info included in the addDriver call though, since it's only an error if the overlap occurs with the same target net? Like you can alias the same range of bits from one net to multiple different nets, it's only error if it's the same nets each time. This example from the LRM I think won't pass with the current code?
```
module overlap(inout wire [15:0] bus16, inout wire [11:0] low12, high12);
alias bus16[11:0] = low12;
alias bus16[15:4] = high12;
endmodule

module overlap(inout wire [15:0] bus16, inout wire [11:0] low12, high12);
alias bus16 = {high12, low12[3:0]};
alias high12[7:0] = low12[11:4];
endmodule
```
Or maybe it will, but not if you reverse the ordering in the alias statements.

File: source/ast/symbols/MemberSymbols.cpp (hunk 8 above)
Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: solution was reimplemented by focusing on merge net alias expressions 

`aliasedSyms` map logic was reduced from `NetAliasVisitor`

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: My solution no longer ignores the leftoutermost net alias expression

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: To relate net alias expressions to existing alias sybmols, I used a comparison of their syntaxes

it's not a best solution but in this case it is possible, since in aliases all symbols are global nets from the point of view of the module, in the future we need to come up with a better comparator

**Comment by likeamahoney** [comment]
closes https://github.com/MikePopoloski/slang/issues/947

File: source/ast/symbols/MemberSymbols.cpp (hunk 8 above)
Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: > I'm not following why the first alias expression should be exempt from the duplicate checking. There isn't anything special about the first alias vs any others; you can reverse the order without changing the behavior.
> 
> I do think we need more info included in the addDriver call though, since it's only an error if the overlap occurs with the same target net? Like you can alias the same range of bits from one net to multiple different nets, it's only error if it's the same nets each time. This example from the LRM I think won't pass with the current code?
> 
> module overlap(inout wire [15:0] bus16, inout wire [11:0] low12, high12);
> alias bus16[11:0] = low12;
> alias bus16[15:4] = high12;
> endmodule
> 
> module overlap(inout wire [15:0] bus16, inout wire [11:0] low12, high12);
> alias bus16 = {high12, low12[3:0]};
> alias high12[7:0] = low12[11:4];
> endmodule
> Or maybe it will, but not if you reverse the ordering in the alias statements.

Provided example works correctly regardless of the alias permutations.

File: source/ast/symbols/MemberSymbols.cpp [hunk 9]
```diff
@@ -2119,18 +2119,40 @@ std::span<const Expression* const> NetAliasSymbol::getNetReferences() const {
...
+    auto& comp = context.getCompilation();
+    auto& aliasedSyms = comp.getNetAliases();
     NetAliasVisitor visitor(context);
 
-    for (auto exprSyntax : syntax->as<NetAliasSyntax>().nets) {
+    const auto nets = syntax->as<NetAliasSyntax>().nets;
+
+    // Preprocessing net alias expressions to relate them to existing net aliases.
+    const NetAliasSymbol* current = this;
+    for (auto exprSyntax : nets) {
```
Comment (MikePopoloski) [inline source/ast/symbols/MemberSymbols.cpp:2134]: The logic here still seems wrong. Consider this simple example:
```
module m;
    wire [1:0] a, b, c;
    alias a = b[1:0];
    alias c = b[1:0];
endmodule
```
which results in the following error (when it should not):
```
source:5:15: error: it is not allowed to specify an alias from an individual signal to itself or to specify a given alias more than once
    alias c = b[1:0];
              ^~~~~~
source:4:15: note: alias declared here
    alias a = b[1:0];
              ^~~~~~
```

I don't think this mergeOrAddNetAlias idea works or is even necessary really. We don't need a compilation-wide map. Each net can already track which bits have been aliased using the driver mechanism. I think what you need to do is to do an addDriver() call across the cartesian product of the aliases in a given net alias statement. Instead of the "containing symbol" being the net alias, instead you make it the net it's being *aliased to*.

As an example:
```
alias a = b = c;
```
you want the following sequence of addDriver calls to be made, where the argument here is what gets set as the "containing symbol" inside ValueDriver:
a.addDriver(b);
a.addDriver(c);
b.addDriver(a);
b.addDriver(c);
c.addDriver(a);
c.addDriver(b);

And then I think the existing logic will just do the right thing without the need for a global map or comparing syntax nodes to each other.

File: include/slang/ast/SemanticFacts.h [hunk 10]
```diff
@@ -132,9 +132,12 @@ enum class SLANG_EXPORT AssignFlags : uint8_t {
 
     /// The assignment is for an output port that was sliced due to an array of instances
     /// being connected to an array argument.
-    SlicedPort = 1 << 7
+    SlicedPort = 1 << 7,
+
+    // The assignment declares net alias
```
[the remark this answers is not in the document]
Comment (likeamahoney) [reply include/slang/ast/SemanticFacts.h]: fixed

File: include/slang/ast/ASTContext.h [hunk 11]
```diff
@@ -298,6 +299,9 @@ class SLANG_EXPORT ASTContext {
     /// Various flags that control how AST creation is performed.
     bitmask<ASTFlags> flags;
 
+    /// Placeholder for storing currently processed net alias
+    const NetAliasSymbol* netAlias = nullptr;
```
[the remark this answers is not in the document]
Comment (likeamahoney) [reply include/slang/ast/ASTContext.h]: removed

File: source/ast/symbols/MemberSymbols.cpp (hunk 9 above)
Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: sorry for delay

solution was reimplemented according to your suggestion

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: but i need to add more complex tests

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: and fix CI

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: CI was fixed and new tests were added

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: CI was fixed and new tests were added

Comment (likeamahoney) [reply source/ast/symbols/MemberSymbols.cpp]: CI was fixed and new tests were added
